rclpy/test/test_time_source.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+import unittest
+
+import builtin_interfaces.msg
+import rclpy
+from rclpy.clock import Clock
+from rclpy.clock import ClockType
+from rclpy.clock import ROSClock
+from rclpy.time import Time
+from rclpy.time_source import CLOCK_TOPIC
+from rclpy.time_source import TimeSource
+
+
+class TestTimeSource(unittest.TestCase):
+
+    def setUp(self):
+        rclpy.init()
+        self.node = rclpy.create_node('TestTimeSource', namespace='/rclpy')
+
+    def tearDown(self):
+        self.node.destroy_node()
+        rclpy.shutdown()
+
+    def publish_clock_messages(self):
+        clock_pub = self.node.create_publisher(builtin_interfaces.msg.Time, CLOCK_TOPIC)
+        cycle_count = 0
+        time_msg = builtin_interfaces.msg.Time()
+        while rclpy.ok() and cycle_count < 5:
+            time_msg.sec = cycle_count
+            clock_pub.publish(time_msg)
+            cycle_count += 1
+            rclpy.spin_once(self.node, timeout_sec=1)
+            # TODO(dhood): use rate once available
+            time.sleep(1)
+
+    def test_time_source_attach_clock(self):
+        time_source = TimeSource(node=self.node)
+
+        # ROSClock is a specialization of Clock with ROS time methods.
+        time_source.attach_clock(ROSClock())
+
+        # A clock of type ROS_TIME can be attached. It will be converted to a ROSClock for storage.
+        time_source.attach_clock(Clock(clock_type=ClockType.ROS_TIME))
+
+        assert all((isinstance(clock, ROSClock) for clock in time_source._associated_clocks))
+
+        with self.assertRaises(ValueError):
+            time_source.attach_clock(Clock(clock_type=ClockType.SYSTEM_TIME))
+
+        with self.assertRaises(ValueError):
+            time_source.attach_clock(Clock(clock_type=ClockType.STEADY_TIME))
+
+    def test_time_source_not_using_sim_time(self):
+        time_source = TimeSource(node=self.node)
+        clock = ROSClock()
+        time_source.attach_clock(clock)
+
+        # When not using sim time, ROS time should look like system time
+        now = clock.now()
+        system_now = Clock(clock_type=ClockType.SYSTEM_TIME).now()
+        assert (system_now.nanoseconds - now.nanoseconds) < 1e9
```

Comment:
`1e9` seems really small. I think this test could be flaky if the OS happens to context switch between lines 73 and 74.

Also, maybe add `abs()` in case there is a future bug where `now` is a non-zero time greater than `system_now`?

Reply:
I was thinking 1sec is pretty generous already but it can't hurt to increase.

About the abs, I don't think it makes sense here, since we _do_ expect the value of a later call to now() be greater. I'd rather catch that bug than mask it.

Reply:
Ooops, I got this completely wrong. For some reason I thought `1e9` meant 1ns. My bad.